### PR TITLE
Separate Object from VisualObject and ContainerObject

### PR DESCRIPTION
**Pull request by rodrigowbazevedo**
Add some interfaces to have the possibility to make abstractions, and separate objects into plain, visual and container.

### CODE REVIEW

File: src/Components/VisualObject.php [hunk 1]
```diff
@@ -0,0 +1,250 @@
+<?php
+
+namespace Gui\Components;
+
+use Gui\Color;
+
+/**
+ * It is a visual component
+ *
+ * @author Gabriel Couto @gabrielrcouto
```
Comment (reisraff) [inline src/Components/VisualObject.php]: the author is incorrect

File: src/Components/VisualObjectInterface.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+namespace Gui\Components;
+
```
Comment (reisraff) [inline src/Components/VisualObjectInterface.php]: Is missing interface summary

File: src/Components/VisualObjectInterface.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+interface VisualObjectInterface
+{
+    public function getAutoSize();
```
Comment (reisraff) [inline src/Components/VisualObjectInterface.php]: as it is an interface, you can just put the annotations over here, and in the implementation you just use @inheritdoc for all methods.

File: src/Components/ContainerObjectInterface.php [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace Gui\Components;
+
```
Comment (reisraff) [inline src/Components/ContainerObjectInterface.php]: Is missing interface summary

File: src/Components/ContainerObjectInterface.php [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+interface ContainerObjectInterface extends LazarusObjectInterface
+{
+    public function appendChild(VisualObjectInterface  $object);
```
Comment (reisraff) [inline src/Components/ContainerObjectInterface.php]: Put the summary here, and in the implementation use @inheritdoc for all methods.

File: src/Components/LazarusObjectInterface.php [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace Gui\Components;
+
```
Comment (reisraff) [inline src/Components/LazarusObjectInterface.php]: Interface summary again

File: src/Components/LazarusObjectInterface.php [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+interface LazarusObjectInterface
+{
+    public function getLazarusObjectId();
```
Comment (reisraff) [inline src/Components/LazarusObjectInterface.php]: methods summary again.

**Comment by reisraff** [comment]
looks good!

**Comment by gabrielrcouto** [comment]
Well done @rodrigowbazevedo! Is it tested? All the examples are working? This will be a big merge :smile:

**Comment by rodrigowbazevedo** [comment]
All the examples were tested @gabrielrcouto!

**Comment by reisraff** [comment]
@gabrielrcouto @rodrigowbazevedo, I guess in this PR can be included unit tests also.

@rodrigowbazevedo for appendChild, getChild, getChildren you can do normally. But for setters/getters you can stub some methods like Object::get Object::set and somethimes Object:call and Application::waitCommand.

**Comment by rodrigowbazevedo** [comment]
@reisraff I'll do it.

**Pull request merged**